src/graph/executor/algo/AllPathsExecutor.cpp

```diff
@@ -292,71 +296,106 @@ folly::Future<Status> AllPathsExecutor::buildResult() {
-      return folly::makeFuture<Status>(Status::OK());
-    }
-    return getPathProps();
-  });
+  return future.via(runner())
+      .ensure([this, buildPathTime]() {
+        otherStats_.emplace("build_path_time",
+                            folly::sformat("{}(us)", buildPathTime.elapsedInUSec()));
+      })
+      .thenValue([this](auto&& resp) {
+        UNUSED(resp);
+        if (!withProp_ || emptyPropVids_.empty()) {
+          finish(ResultBuilder().value(Value(std::move(result_))).build());
+          return folly::makeFuture<Status>(Status::OK());
+        }
+        return getPathProps();
+      });
 }
 
 folly::Future<Status> AllPathsExecutor::buildPathMultiJobs() {
-  auto pathsPtr = std::make_shared<std::vector<std::vector<Value>>>();
+  auto pathsPtr = std::make_shared<std::vector<NPath*>>();
+  if (threadLocalPtr_.get() == nullptr) {
+    threadLocalPtr_.reset(new std::deque<NPath>());
+  }
   for (auto& vid : leftInitVids_) {
     auto vidIter = leftAdjList_.find(vid);
     if (vidIter == leftAdjList_.end()) {
       continue;
     }
-    auto src = vidIter->first;
+    auto& src = vidIter->first;
     auto& adjEdges = vidIter->second;
     if (adjEdges.empty()) {
       continue;
     }
     pathsPtr->reserve(adjEdges.size() + pathsPtr->size());
     for (auto& edge : adjEdges) {
-      pathsPtr->emplace_back(std::vector<Value>({src, edge}));
+      threadLocalPtr_->emplace_back(NPath(src, edge));
+      pathsPtr->emplace_back(&threadLocalPtr_->back());
     }
   }
   size_t step = 2;
   auto future = doBuildPath(step, 0, pathsPtr->size(), pathsPtr);
-  return future.via(runner()).thenValue([this](std::vector<Row>&& paths) {
+  return future.via(runner()).thenValue([this](std::vector<std::pair<NPath*, Value>>&& paths) {
     memory::MemoryCheckGuard guard;
+
     if (!paths.empty()) {
-      result_.rows.swap(paths);
+      time::Duration convertPathTime;
+      for (auto& path : paths) {
+        result_.rows.emplace_back(convertNPath2Row(path.first, path.second));
+      }
+      otherStats_.emplace("convert_path_time",
+                          folly::sformat("{}(us)", convertPathTime.elapsedInUSec()));
     }
     return Status::OK();
   });
 }
 
-folly::Future<std::vector<Row>> AllPathsExecutor::doBuildPath(
-    size_t step,
-    size_t start,
-    size_t end,
-    std::shared_ptr<std::vector<std::vector<Value>>> pathsPtr) {
-  if (cnt_.load(std::memory_order_relaxed) >= limit_) {
-    return folly::makeFuture<std::vector<Row>>(std::vector<Row>());
+// construct ROW[src1, [e1, v2, e2], v3]
+Row AllPathsExecutor::convertNPath2Row(NPath* path, Value dst) {
+  std::vector<Value> list;
+  NPath* head = path;
+  while (head != nullptr) {
+    list.emplace_back(head->edge);
+    list.emplace_back(head->vertex);
+    head = head->p;
   }
+  Row row;
+  // add src;
+  row.values.emplace_back(list.back());
+  list.pop_back();
+  std::reverse(list.begin(), list.end());
```

Comment:
why need to reverse the list?

Reply:
when recursively parsing points and edges through NPath, the order is reversed compared to the topology of the path